File: src/lorekeeper/dashboard/routes/suggestions.py
"""Dashboard API routes for suggestion review and sweep control."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel

from lorekeeper.dashboard.handler import DashboardHandler

router = APIRouter()


# ── Request / Response schemas ────────────────────────────────────────────────


class BatchAction(BaseModel):
    suggestion_ids: list[str]
    action: str  # "accept" | "reject"


class BatchResultItem(BaseModel):
    id: str
    status: str  # "accepted" | "rejected" | "error"
    message: str


class BatchResponse(BaseModel):
    results: list[BatchResultItem]
    accepted: int = 0
    rejected: int = 0
    errors: list[str] = []


def _handler(request: Request) -> DashboardHandler:
    return request.app.state.dashboard_handler  # type: ignore[no-any-return]


# ── Routes ────────────────────────────────────────────────────────────────────


@router.get("/api/suggestions")
def list_suggestions(
    request: Request,
    limit: int = 50,
    offset: int = 0,
    sort_by: str = "weighted_score",
    sort_dir: str = "desc",
    memory_id: str | None = None,
) -> dict[str, Any]:
    """List pending suggestions with pagination."""
    try:
        return _handler(request).list_suggestions(
            limit=limit, offset=offset, sort_by=sort_by, sort_dir=sort_dir, memory_id=memory_id,
        )
    except ValueError as e:
        raise HTTPException(status_code=422, detail=str(e)) from e


@router.get("/api/suggestions/count")
def count_suggestions(request: Request, memory_id: str | None = None) -> dict[str, int]:
    """Return total pending suggestion count, optionally filtered by memory."""
    return _handler(request).count_suggestions(memory_id=memory_id)


@router.post("/api/suggestions/batch")
def batch_suggestions(request: Request, body: BatchAction) -> BatchResponse:
    """Accept or reject multiple suggestions at once."""
    action = body.action.lower()
    if action not in ("accept", "reject"):
        raise HTTPException(
            status_code=422,
            detail="action must be 'accept' or 'reject'",
        )

    result = _handler(request).batch_suggestions(
        suggestion_ids=body.suggestion_ids, action=action
    )

    mapped_results = [
        BatchResultItem(id=r["id"], status=r["status"], message=r["message"])
        for r in result["results"]
    ]

    return BatchResponse(
        results=mapped_results,
        accepted=result["accepted"],
        rejected=result["rejected"],
        errors=result["errors"],
    )


@router.post("/api/sweep/trigger")
def trigger_sweep(request: Request) -> dict[str, bool]:
    """Trigger the sweep scheduler immediately by resetting its timer."""
    return _handler(request).trigger_sweep()


@router.get("/api/sweep/status")
def sweep_status(request: Request) -> dict[str, str | None]:
    """Return last and next sweep run timestamps."""
    return _handler(request).sweep_status()

File: src/dashboard_v2/src/lib/api/suggestions.ts
/**
 * API helpers for the review inbox (suggestions).
 */

const BASE = '';

async function api<T>(method: string, path: string, body?: unknown): Promise<T> {
	const res = await fetch(`${BASE}${path}`, {
		method,
		headers: body ? { 'Content-Type': 'application/json' } : undefined,
		body: body ? JSON.stringify(body) : undefined,
	});
	if (!res.ok) {
		throw new Error(`API ${method} ${path}: ${res.status} ${res.statusText}`);
	}
	return res.json() as Promise<T>;
}

export interface SuggestionRow {
	id: string;
	source_memory_id: string;
	source_title: string;
	target_memory_id: string;
	target_title: string;
	weighted_score: number;
	cosine_score: number;
	bm25_score: number;
	entity_score: number;
	temporal_score: number;
	confidence: number;
	created_at: string;
}

export interface SuggestionsResponse {
	items: SuggestionRow[];
	total: number;
	offset: number;
}

export interface SuggestionsParams {
	limit?: number;
	offset?: number;
	sort_by?: string;
	sort_dir?: 'asc' | 'desc';
	memory_id?: string;
	status?: 'pending' | 'reviewed';
}

export interface BatchResultItem {
	id: string;
	status: string;
	message: string;
}

export interface BatchResponse {
	results: BatchResultItem[];
	accepted: number;
	rejected: number;
	errors: string[];
}

function buildQuery(params: SuggestionsParams): string {
	const p = new URLSearchParams();
	if (params.limit !== undefined) p.set('limit', String(params.limit));
	if (params.offset !== undefined) p.set('offset', String(params.offset));
	if (params.sort_by) p.set('sort_by', params.sort_by);
	if (params.sort_dir) p.set('sort_dir', params.sort_dir);
	if (params.memory_id) p.set('memory_id', params.memory_id);
	if (params.status) p.set('status', params.status);
	const s = p.toString();
	return s ? `?${s}` : '';
}

export function fetchSuggestions(params: SuggestionsParams = {}): Promise<SuggestionsResponse> {
	return api<SuggestionsResponse>('GET', `/api/suggestions${buildQuery(params)}`);
}

export function fetchSuggestionCount(memory_id?: string): Promise<{ count: number }> {
	const qs = memory_id ? `?memory_id=${encodeURIComponent(memory_id)}` : '';
	return api<{ count: number }>('GET', `/api/suggestions/count${qs}`);
}

export function batchSuggestions(
	suggestion_ids: string[],
	action: 'accept' | 'reject',
): Promise<BatchResponse> {
	return api<BatchResponse>('POST', '/api/suggestions/batch', { suggestion_ids, action });
}
